flink-table/flink-table-common/src/test/java/org/apache/flink/table/architecture/ArchitectureTests.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.architecture;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.Public;
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.table.factories.DynamicTableFactory;
+import org.apache.flink.testutils.architecture.ApiRules;
+
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaField;
+import com.tngtech.archunit.core.domain.JavaModifier;
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.junit.ArchTests;
+import com.tngtech.archunit.junit.ArchUnitRunner;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import org.junit.runner.RunWith;
+
+import java.util.Set;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static org.apache.flink.testutils.architecture.Predicates.areAnnotatedWithAtLeastOneOf;
+import static org.apache.flink.testutils.architecture.Predicates.fulfill;
+
+/** TODO. */
+@RunWith(ArchUnitRunner.class)
+@AnalyzeClasses(
+        packages = "org.apache.flink.table",
+        importOptions = ImportOption.DoNotIncludeTests.class)
+public class ArchitectureTests {
+    @ArchTest public static final ArchTests APIS = ArchTests.in(ApiRules.class);
+
+    /** Ensure that all classes in the API package have a proper annotation. */
+    @ArchTest
+    public static final ArchRule ANNOTATED_APIS =
```

Comment:
In flink-table-common this test currently identifies a couple of classes (all exceptions) which are not annotated.